pgmax/fg/groups.py

```diff
@@ -586,27 +597,52 @@ def __getitem__(self, variables: Union[Sequence, Collection]) -> nodes.Factor:
             ValueError: if the queried factor is not present in the factor group
         """
         variables = frozenset(variables)
-        if variables not in self._variables_to_factors:
+        if variables not in self._variables_to_factors():
             raise ValueError(
                 f"The queried factor connected to the set of variables {variables} is not present in the factor group."
             )
 
         return self._variables_to_factors[variables]
 
+    @cached_property
+    def factors(self) -> Tuple[nodes.Factor, ...]:
+        """Returns all factors in the factor group.
+        This function is only called on demand when the user requires it."""
+        return tuple(self._variables_to_factors.values())
+
+    @cached_property
+    def _variables_to_factors(self) -> Mapping[FrozenSet, nodes.Factor]:
+        """Function to compile potential array for the factor group.
+        This function is only called on demand when the user requires it.
+
+        Returns:
+            A dictionnary mapping set of connected variables to the corresponding factors
+        """
+        return self._get_variables_to_factors()
+
     @cached_property
     def factor_group_log_potentials(self) -> np.ndarray:
+        """Flattened array of log potentials"""
+        return self.factor_group_log_potentials_full.flatten()
+
+    @cached_property
+    def factor_group_log_potentials_full(self) -> np.ndarray:
```

Comment:
Why is there a full version? If this is the non-flat version it should just be used to define the class, and shouldn't be a property